File: solutions/java/src/digitalwallet/DigitalWallet.java
package digitalwallet;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class DigitalWallet {
    private static DigitalWallet instance;
    private final Map<String, User> users;
    private final Map<String, Account> accounts;
    private final Map<String, PaymentMethod> paymentMethods;

    private DigitalWallet() {
        users = new ConcurrentHashMap<>();
        accounts = new ConcurrentHashMap<>();
        paymentMethods = new ConcurrentHashMap<>();
    }

    public static synchronized DigitalWallet getInstance() {
        if (instance == null) {
            instance = new DigitalWallet();
        }
        return instance;
    }

    public void createUser(User user) {
        users.put(user.getId(), user);
    }

    public User getUser(String userId) {
        return users.get(userId);
    }

    public void createAccount(Account account) {
        accounts.put(account.getId(), account);
        account.getUser().addAccount(account);
    }

    public Account getAccount(String accountId) {
        return accounts.get(accountId);
    }

    public void addPaymentMethod(PaymentMethod paymentMethod) {
        paymentMethods.put(paymentMethod.getId(), paymentMethod);
    }

    public PaymentMethod getPaymentMethod(String paymentMethodId) {
        return paymentMethods.get(paymentMethodId);
    }

    public synchronized void transferFunds(Account sourceAccount, Account destinationAccount, BigDecimal amount, Currency currency) {
        if (sourceAccount.getCurrency() != currency) {
            amount = CurrencyConverter.convert(amount, currency, sourceAccount.getCurrency());
        }
        sourceAccount.withdraw(amount);

        if (destinationAccount.getCurrency() != currency) {
            amount = CurrencyConverter.convert(amount, currency, destinationAccount.getCurrency());
        }
        destinationAccount.deposit(amount);

        String transactionId = generateTransactionId();
        Transaction transaction = new Transaction(transactionId, sourceAccount, destinationAccount, amount, currency);
        sourceAccount.addTransaction(transaction);
        destinationAccount.addTransaction(transaction);
    }

    public List<Transaction> getTransactionHistory(Account account) {
        return account.getTransactions();
    }

    private String generateTransactionId() {
        return "TXN" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
    }
}

File: solutions/golang/digitalwalletservice/digital_wallet.go
package digitalwallet

import (
	"fmt"
	"math/big"
	"sync"
	"time"
)

type DigitalWallet struct {
	users          map[string]*User
	accounts       map[string]*Account
	paymentMethods map[string]PaymentMethod
	mu             sync.RWMutex
}

var (
	instance *DigitalWallet
	once     sync.Once
)

func GetDigitalWallet() *DigitalWallet {
	once.Do(func() {
		instance = &DigitalWallet{
			users:          make(map[string]*User),
			accounts:       make(map[string]*Account),
			paymentMethods: make(map[string]PaymentMethod),
		}
	})
	return instance
}

func (dw *DigitalWallet) CreateUser(user *User) {
	dw.mu.Lock()
	defer dw.mu.Unlock()
	dw.users[user.ID] = user
}

func (dw *DigitalWallet) CreateAccount(account *Account) {
	dw.mu.Lock()
	defer dw.mu.Unlock()
	dw.accounts[account.ID] = account
	account.User.AddAccount(account)
}

func (dw *DigitalWallet) AddPaymentMethod(method PaymentMethod) {
	dw.mu.Lock()
	defer dw.mu.Unlock()
	dw.paymentMethods[method.GetID()] = method
}

func (dw *DigitalWallet) TransferFunds(sourceAccount, destinationAccount *Account, amount *big.Float, currency Currency) error {
	dw.mu.Lock()
	defer dw.mu.Unlock()

	converter := GetCurrencyConverter()

	// Convert amount to source account currency
	sourceAmount := amount
	if sourceAccount.Currency != currency {
		sourceAmount = converter.Convert(amount, currency, sourceAccount.Currency)
	}

	// Withdraw from source account
	if err := sourceAccount.Withdraw(sourceAmount); err != nil {
		return err
	}

	// Convert amount to destination account currency
	destAmount := amount
	if destinationAccount.Currency != currency {
		destAmount = converter.Convert(amount, currency, destinationAccount.Currency)
	}

	// Deposit to destination account
	destinationAccount.Deposit(destAmount)

	// Create and record transaction
	transactionID := fmt.Sprintf("TXN%d", time.Now().UnixNano())
	transaction := NewTransaction(transactionID, sourceAccount, destinationAccount, amount, currency)
	sourceAccount.AddTransaction(transaction)
	destinationAccount.AddTransaction(transaction)

	return nil
}

func (dw *DigitalWallet) GetTransactionHistory(account *Account) []*Transaction {
	return account.GetTransactions()
}
